Rewrite pool2d fwd

jakpiase opened the pull request:
This PR rewrites pool2d so it doesn't use pool3d with 1's inserted as D dims in NCDHW

jakpiase committed:
added pool2d fwd

bartekxk commented on library/src/tensor_operation_instance/gpu/pool2d_fwd/pool2d_fwd_instance_common.hpp:
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: MIT
```
Can we move it to library/include/ck/library/tensor_operation_instance/gpu/pool2d_fwd ?

bartekxk reviewed (commented):
LGTM

bartekxk reviewed (commented):
Please extend 3d pool tests to run also 2d cases

jakpiase committed:
add tests

jakpiase committed:
Merge remote-tracking branch 'origin/develop' into jakpiase/pool2d_fwd_new

jakpiase committed:
Merge branch 'develop' into jakpiase/pool2d_fwd_new

mozga-amd commented on profiler/src/profile_max_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,239 @@
+    constexpr auto ReduceOpId = ck::ReduceTensorOp::AVG;
+#endif
+
+    if(false)
```
What this condition is used for?

jakpiase replied:
It's a weird construction to allow nested #ifdefs and `else` at the end of if-else block. Without it the code won't compile for some cases in which certain datatypes are not supported

aosewski replied:
I wonder whether you really need here all those `#ifdef CK_ENABLE_xxxx`s.... If you would get rid of them here I think it should work fine. You won't get compiled unwanted instances anyway, since you have those macros already in here: library/include/ck/library/tensor_operation_instance/gpu/pool2d_fwd.hpp

mozga-amd commented on profiler/src/profile_max_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,239 @@
+#ifdef CK_ENABLE_BF16
+    else if(data_type == ck::DataTypeEnum::BFloat16)
+    {
+        if(return_index)
```
At the first glance, the fn-calls differ by only one template-value `true\false`. So, can't we just make one call and pass `return_index` as parameter?

jakpiase replied:
`return_index` is not known at compile time so we can't do that since we must support runtime parameters. The only way is to use the ugly if/else

mozga-amd replied:
Oh yes, I see it now, it's also being retrieved from the command line. I missed that line out.

mozga-amd commented on test/pool/test_avg_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,65 @@
+
+#ifdef CK_ENABLE_FP16
+using KernelTypes =
+    ::testing::Types<std::tuple<F16, F16, F32, I32>, std::tuple<F32, F32, F32, I32>>;
```
In the profiler, we have input `F16,F16,F16,` for `CK_ENABLE_FP16` enabled, here is `F16,F16,F32`. Is this intentional?

jakpiase replied:
That's a great catch. The code was initially the same as for pool3d and I haven't seen this detail. I believe that `F16,F16,F16` is the correct one, if not for both then at least for max pooling since there are no numerical calculations so it may be slightly faster but in terms of correct results there is absolutely no difference whether the third parameter is `F16` or `F32`.

mozga-amd commented on test/pool/test_max_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,86 @@
+
+#ifdef CK_ENABLE_FP16
+using KernelTypes =
+    ::testing::Types<std::tuple<F16, F16, F32, I32>, std::tuple<F32, F32, F32, I32>>;
```
Same here.. In the profiler, we have input `F16,F16,F16`, but for `CK_ENABLE_FP16` enabled, here is `F16,F16,F32`. Is this intentional?

mozga-amd commented on test/pool/test_max_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,86 @@
+        for(auto param : params)
+        {
+            // max pool
+            bool success =
```
It’s worth considering simplifying this by splitting this test it into two tests, one with an index and one without, and passing the `index` as a parameter to the runner.

aosewski replied:
Agree with Mateusz. I'd add output index as a additional tparam and have two separate test cases.

mozga-amd replied:
Remember that new test, creates new `this->params`. To avoid that and speed up UTs y can use: `SetUpTestCase()`

mozga-amd commented on profiler/src/profile_max_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,239 @@
+    else if(data_type == ck::DataTypeEnum::BFloat16)
+    {
+        if(return_index)
+            ck::profiler::
```
Is it possible to simplify this syntax and make just one call to `profile_pool2d_fwd_impl`? The template parameters can be standardized and set at the top as we do: instead of `BF16, F16`— we can name them more generally. Here, you would only need to assign values.

jakpiase replied:
There are two calls because there is a parameter which is not known at compile time. How would you like to refactor it to use only one call? Maybe there is a way that I don't see now

mozga-amd replied:
By one call, I mean removing the `#ifdef` statements that define the type and controlling the execution through one if. The types are defined in lines 112-121.
```cpp
if (data_type != ck::DataTypeEnum::Half...) {
  throw std::runtime : error("Not implemented yet");
  return 0;
}
if (return_index) {
  // call: with index
   profile_pool2d_fwd_impl<TypeA, TypeA, TypeA, TypeB, NHWC, NHWC, ReduceOpId, false, true>
} else {
  // call: no index
}
```

aosewski replied:
No it's not possible since `data_type` is a variable known only at runtime.

mozga-amd commented on profiler/src/profile_max_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,239 @@
+#ifdef CK_ENABLE_FP16
+    else if(data_type == ck::DataTypeEnum::Half)
+    {
+        if(return_index)
```
[Consider] It could be fine to use bracket in one-line if-else statement

aosewski reviewed (changes requested):
Please split test cases with respect to output index value. Moreover consider removing CK_ENABLE_ macros from profiler source files.

mozga-amd commented on test/pool/test_avg_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,65 @@
+    }
+};
+
+#ifdef CK_ENABLE_FP16
```
To consider: To avoid `#ifdef` condition it could be also replaced with: 
```cpp
using KernelTypes = typename std::conditioanl<CK_ENABLE_FP16>::template type<
    ::testing::Types<std::tuple<F16, F16, F32, I32>,
                     std::tuple<F32, F32, F32, I32>>,
    ::testing::Types<std::tuple<F32, F32, F32, I32>>>
```
or with `std::conditional_t`

jakpiase committed:
add reviewers changes

jakpiase committed:
Merge remote-tracking branch 'origin/develop' into jakpiase/pool2d_fwd_new

jakpiase commented:
Added reviewers requests and changed ops names to enable alphabetical search between different types of pooling since now avg_pool and max_pool were not close to each other in file list

aosewski commented on test/pool/test_pool2d_max_fwd.cpp:
```diff
@@ -30,48 +31,28 @@ class TestMaxPool2dFwd : public ::testing::Test
-    ::testing::Types<std::tuple<F16, F16, F32, I32>, std::tuple<F32, F32, F32, I32>>;
-#else
-using KernelTypes = ::testing::Types<std::tuple<F32, F32, F32, I32>>;
-#endif
```
I wonder whether this change can cause CI to fail. If There will be no instance found (in case when specific type is not defined) then profiler will return false and overall test case will fail.... However the ifdef..else chain as below in one of the following files is ugly. Think about if we can have more elegant solution? Current CMake can skip compilation of files which does not match to defined data types, but splitting tests into multiple files based on data type isn't better solution anyway.

aosewski commented on include/ck/tensor_operation/gpu/device/impl/device_pool2d_fwd_nhwc_nhwc_new.hpp:
This file name is awkward :) What about `_v2` ?? Or even what about replacing the old version with this one entirely if this version is faster? Anyway if this version isn't faster then there's no point in having it. Have you profiled this ?

aosewski reviewed (commented):
I have two concerns. Moreover I'm not sure about this renaming, since I believe the DL community standard of naming pooling ops is with the reduction op name as prefix ie.: "avg_pool". Thus I wonder if we won't make more confusion with this, however I know this is just the file names and not the operator names, which are more important.

mozga-amd commented on profiler/src/profile_pool2d_max_fwd.cpp:
```diff
@@ -0,0 +1,233 @@
+              << "arg4: print tensor value (0: no; 1: yes)\n"
+              << "arg5: time kernel (0=no, 1=yes)\n"
+              << "arg6: return index (0=no, 1=yes)\n"
+              << "--length: input tensor length for NCDHW(e.g, --length 2 32 30 30 30) \n"
```
It looks like that description is out-of-date: NCDHW, strides etc.

mozga-amd commented on include/ck/tensor_operation/gpu/device/impl/device_pool2d_fwd_nhwc_nhwc.hpp:
```diff
@@ -16,95 +27,363 @@ template <typename InDataType,
-    using DevicePool3D = DevicePool3dFwd_NDHWC_NDHWC<InDataType,
+    static constexpr auto I0 = Number<0>{};
+    static constexpr auto I1 = Number<1>{};
+    static constexpr auto I2 = Number<2>{};
```
It looks like I2-I5 are not used.

jakpiase committed:
Revert "Merge remote-tracking branch 'origin/develop' into jakpiase/pool2d_fwd_new"

This reverts commit 6b2ba7ff8960b0a6ddbe30d8dac53eeb55a8597e, reversing
changes made to 22c82bea0caf3e0f29399100c1bb67b8003fc042.

jakpiase committed:
Revert "add reviewers changes"

This reverts commit 22c82bea0caf3e0f29399100c1bb67b8003fc042.

jakpiase committed:
added reviewers comments

jakpiase committed:
revert some old files

jakpiase committed:
Merge remote-tracking branch 'origin/develop' into jakpiase/pool2d_fwd_new

aosewski commented on profiler/src/profile_max_pool2d_fwd.cpp:
```diff
@@ -109,18 +109,17 @@ int profile_pool2d_max_fwd(int argc, char* argv[])
     {
-        if(return_index)
-        {
+        if(return_index) {
```
Please run clang-format. We used to put braces in new line.

aosewski commented on profiler/include/profiler/profile_pool2d_fwd_impl.hpp:
```diff
@@ -0,0 +1,274 @@
+    out_length[0] = N;
+    out_length[1] = C;
+
+    // Calculate Do, Ho, Wo
```
Please update comment.

aosewski commented on test/pool/test_max_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,72 @@
+                     std::tuple<F32, F32, F32, I32, false_t>>,
+    ::testing::Types<std::tuple<F32, F32, F32, I32, true_t>,
+                     std::tuple<F32, F32, F32, I32, false_t>>
+>;
```
You do not test here your BF16 instances. Please add them.

aosewski commented on test/pool/test_avg_pool2d_fwd.cpp:
```diff
@@ -0,0 +1,68 @@
+                     std::tuple<F32, F32, F32, I32>>,
+    ::testing::Types<std::tuple<F32, F32, F32, I32>,
+                     std::tuple<F32, F32, F32, I32>>
+>;
```
You do not test here your BF16 instances. Please add them.

aosewski reviewed (changes requested):
Few things to fix:
* Please run one more time clang format on all files: I see non-standard formatting.
* Please add test cases for BF16 instances max & avg

aosewski commented:
@jakpiase What about performance ? Is it faster ?

jakpiase committed:
add reviewers requests

jakpiase committed:
Merge remote-tracking branch 'origin/develop' into jakpiase/pool2d_fwd_new

jakpiase committed:
Merge branch 'develop' into jakpiase/pool2d_fwd_new

aosewski committed:
Merge branch 'develop' into jakpiase/pool2d_fwd_new
